File: agents/profit_optimizer.py
"""Always plants whichever affordable crop has the highest expected profit
per slot per day, and reinvests aggressively. Purpose: surface dominant
crops and upgrade exploits.
"""

from agents.base import Agent
from simulation import contracts, economy_rules, inventory


class ProfitOptimizer(Agent):
    name = "profit_optimizer"
    description = "Maximizes expected profit per slot per day; waters reliably; fertilizes only when the math says it pays off."

    # Below this soil_health_factor, relax (not remove) the fertilizer
    # marginal-profit bar as maintenance spend -- soil nutrients never
    # regenerate on their own (see economy_rules.soil_quality_risk), so a
    # strict per-crop EV check under-fertilizes crops that are only mildly
    # unprofitable to fertilize once soil health is factored in. This never
    # forces fertilizer onto a crop whose own economics are deeply negative
    # (e.g. quickweed, at roughly -6 nominal marginal profit) -- that would
    # just trade a bankruptcy-by-quality-rejection for a bankruptcy-by-
    # fertilizer-bill.
    SOIL_HEALTH_FERTILIZE_THRESHOLD = 0.5
    SOIL_MAINTENANCE_MARGINAL_PROFIT_FLOOR = -3.0

    def choose_crop(self, player, crops, crops_by_id, upgrades_by_id):
        candidates = [
            c
            for c in crops
            if economy_rules.is_crop_unlocked(c, player) and player.money >= c["seed_cost"]
        ]
        # Never sink a seed cost into a crop that provably can't mature (and
        # so can't be sold) before the run ends -- if that empties the
        # candidate list, leaving the slot idle for the rest of the run is
        # correct, not a bug to fall back around. The boundary itself lives
        # in economy_rules.matures_within_run, shared with the contract
        # forecasts and the other agents that need it.
        candidates = [
            c
            for c in candidates
            if economy_rules.matures_within_run(
                economy_rules.effective_growth_days(c, player, upgrades_by_id), player
            )
        ]
        if not candidates:
            return None
        # `candidates` above already excludes anything unlocked/affordable/
        # unable to mature by the final simulated day, so looking the
        # contracted crop up there (rather than back in crops_by_id) reuses
        # that same maturity check instead of silently bypassing it.
        candidates_by_id = {c["id"]: c for c in candidates}
        active = next(
            (
                contract
                for contract in player.active_contracts
                # Deadline resolution runs at end of day, after crop
                # decisions -- a contract past its deadline but not yet
                # resolved must not still drive today's planting.
                if not contract.resolved and player.day <= contract.deadline_day
            ),
            None,
        )
        if active:
            contracted_crop = candidates_by_id.get(active.item_id)
            if contracted_crop is not None:
                days_to_deadline = active.deadline_day - player.day
                growth_days = economy_rules.effective_growth_days(
                    contracted_crop, player, upgrades_by_id
                )
                still_short = contracts.forecast_committed_supply(player, active) < active.remaining
                if growth_days <= days_to_deadline and still_short:
                    return contracted_crop

        crop = economy_rules.choose_crop_with_relaxed_reserve(
            candidates, player, upgrades_by_id, reserve_fractions=(1.0, 0.5, 0.25)
        )
        if crop is not None:
            return crop
        return min(candidates, key=lambda c: c["growth_days"])

    def should_buy_upgrade(self, player, upgrade):
        return economy_rules.should_buy_upgrade_within_budget(player, upgrade)

    def should_fertilize(self, player, planted, crop, fertilizer_config):
        marginal_profit = economy_rules.fertilizer_expected_marginal_profit(crop, fertilizer_config)
        low_soil_health = (
            economy_rules.soil_health_factor(player) < self.SOIL_HEALTH_FERTILIZE_THRESHOLD
        )
        if low_soil_health and marginal_profit > self.SOIL_MAINTENANCE_MARGINAL_PROFIT_FLOOR:
            # Soil maintenance is exempt from the full reserve (only raw
            # affordability): once soil health is critically low, skipping
            # a cheap, not-badly-unprofitable fertilizer application to
            # protect the reserve just guarantees the next harvest gets
            # rejected outright, which hurts cash flow far more than the
            # fertilizer would.
            return player.money >= fertilizer_config["cost"]
        if not economy_rules.can_spend_with_reserve(player, fertilizer_config["cost"]):
            return False
        return marginal_profit > 0

    def choose_contracts(self, player, offers):
        if any(not contract.resolved for contract in player.active_contracts):
            return []
        suitable = [
            offer
            for offer in offers
            if contracts.is_offer_profitable(player, offer)
            and contracts.is_offer_feasible(player, offer)
        ]
        return [max(suitable, key=lambda offer: offer.unit_price).id] if suitable else []

    def choose_processing(self, player, recipes, items_by_id):
        """Rank recipes by margin per batch and reserve input inventory,
        cash, and processing-capacity slots across the whole plan, so what's
        emitted here is exactly what `simulation.processing.start_job` can
        actually execute -- not an independent per-recipe guess against the
        unchanged starting state that silently drops (or under-batches)
        once capacity/cash/inputs run out. Ties keep `recipes`' own order
        (a stable sort on the margin key alone), so output only depends on
        recipe order when two recipes are exactly equally profitable.
        """
        # processing_capacity is None before the engine has ever set it (e.g.
        # a bare PlayerState in a unit test); treat that the same as 0, per
        # the same convention simulation/contracts.py's _processing_capacity
        # uses, rather than crashing on `None - int`.
        remaining_capacity = (player.processing_capacity or 0) - len(player.processing_jobs)
        if remaining_capacity <= 0:
            return []

        profitable = []
        for recipe in recipes:
            input_value = (
                player.market_prices.get(recipe["input_item_id"], 0) * recipe["input_quantity"]
            )
            output_value = (
                player.market_prices.get(recipe["output_item_id"], 0) * recipe["output_quantity"]
            )
            margin_per_batch = output_value - input_value - recipe.get("cost", 0)
            if margin_per_batch > 0:
                profitable.append((margin_per_batch, recipe))
        profitable.sort(key=lambda scored: -scored[0])

        decisions = []
        reserved_input = {}
        cash_remaining = player.money
        for _margin, recipe in profitable:
            if remaining_capacity <= 0:
                break
            input_item_id = recipe["input_item_id"]
            already_reserved = reserved_input.get(input_item_id, 0)
            available = (
                inventory.available_quantity(
                    player, input_item_id, recipe.get("min_quality", "processing")
                )
                - already_reserved
            )
            max_by_input = available // recipe["input_quantity"]
            cost_per_batch = recipe.get("cost", 0)
            max_by_cash = (
                int(cash_remaining // cost_per_batch) if cost_per_batch > 0 else max_by_input
            )
            batches = min(remaining_capacity, max_by_input, max_by_cash)
            if batches <= 0:
                continue
            decisions.append({"recipe_id": recipe["id"], "batches": batches})
            remaining_capacity -= batches
            reserved_input[input_item_id] = already_reserved + batches * recipe["input_quantity"]
            cash_remaining -= batches * cost_per_batch
        return decisions

    def choose_sales(self, player, channels, items_by_id):
        return self.route_sales_by_best_price(player, channels)

    def should_use_fertilizer(self, player, crop, fertilizer_config):
        marginal_profit = economy_rules.fertilizer_expected_marginal_profit(crop, fertilizer_config)
        low_soil_health = (
            economy_rules.soil_health_factor(player) < self.SOIL_HEALTH_FERTILIZE_THRESHOLD
        )
        if low_soil_health and marginal_profit > self.SOIL_MAINTENANCE_MARGINAL_PROFIT_FLOOR:
            return player.money >= crop["seed_cost"] + fertilizer_config["cost"]
        if not economy_rules.can_spend_with_reserve(
            player, crop["seed_cost"] + fertilizer_config["cost"]
        ):
            return False
        return marginal_profit > 0

File: farm-c/src/agents/profit_optimizer.c
/* Faithful port of agents/profit_optimizer.py. */
#include "agents/profit_optimizer.h"

#include <stdlib.h>

#include "contracts.h"
#include "economy.h"
#include "inventory.h"
#include "vec_util.h"

/* agents/profit_optimizer.py:23-24 */
#define SOIL_HEALTH_FERTILIZE_THRESHOLD 0.5
#define SOIL_MAINTENANCE_MARGINAL_PROFIT_FLOOR (-3.0)

/* agents/profit_optimizer.py:26-79 */
ItemId profit_optimizer_choose_crop(const Agent *self, const FarmState *state,
                                     const ResolvedConfig *config) {
    (void)self;
    if (config->crop_count == 0) {
        return INVALID_ID;
    }
    const CropDef *candidates[config->crop_count];
    size_t candidate_count = 0;
    for (size_t i = 0; i < config->crop_count; i++) {
        const CropDef *crop = &config->crops[i];
        if (economy_is_crop_unlocked(crop, state) && state->money >= crop->seed_cost) {
            candidates[candidate_count++] = crop;
        }
    }
    size_t matures_count = 0;
    for (size_t i = 0; i < candidate_count; i++) {
        int growth_days = economy_effective_growth_days(candidates[i], state, config);
        if (economy_matures_within_run(growth_days, state)) {
            candidates[matures_count++] = candidates[i];
        }
    }
    candidate_count = matures_count;
    if (candidate_count == 0) {
        return INVALID_ID;
    }

    const ContractRecord *active = NULL;
    for (size_t i = 0; i < state->active_contracts.count; i++) {
        const ContractRecord *contract = &state->active_contracts.data[i];
        if (!contract->resolved && state->day <= contract->deadline_day) {
            active = contract;
            break;
        }
    }
    if (active != NULL) {
        const CropDef *contracted_crop = NULL;
        for (size_t i = 0; i < candidate_count; i++) {
            if (candidates[i]->item_id == active->item_id) {
                contracted_crop = candidates[i];
                break;
            }
        }
        if (contracted_crop != NULL) {
            int days_to_deadline = active->deadline_day - state->day;
            int growth_days = economy_effective_growth_days(contracted_crop, state, config);
            bool still_short =
                contracts_forecast_committed_supply(state, config, active) < contract_remaining(active);
            if (growth_days <= days_to_deadline && still_short) {
                return contracted_crop->item_id;
            }
        }
    }

    static const double RESERVE_FRACTIONS[3] = {1.0, 0.5, 0.25};
    const CropDef *crop = economy_choose_crop_with_relaxed_reserve(
        candidates, candidate_count, state, config, RESERVE_FRACTIONS, 3);
    if (crop != NULL) {
        return crop->item_id;
    }

    const CropDef *fastest = candidates[0];
    for (size_t i = 1; i < candidate_count; i++) {
        if (candidates[i]->growth_days < fastest->growth_days) {
            fastest = candidates[i];
        }
    }
    return fastest->item_id;
}

/* agents/profit_optimizer.py:81-82 */
bool profit_optimizer_should_buy_upgrade(const Agent *self, const FarmState *state,
                                          UpgradeId upgrade_id) {
    (void)self;
    const UpgradeDef *upgrade = config_find_upgrade(state->config, upgrade_id);
    return economy_should_buy_upgrade_within_budget(
        state, upgrade, state->config, ECONOMY_UPGRADE_COOLDOWN_DAYS_DEFAULT,
        ECONOMY_UPGRADE_MIN_PAYBACK_MULTIPLE_DEFAULT,
        ECONOMY_UPGRADE_MAX_CUMULATIVE_SPEND_FRACTION_DEFAULT,
        ECONOMY_UPGRADE_DEFAULT_PAYBACK_HORIZON_DAYS_DEFAULT);
}

/* agents/profit_optimizer.py:84-99 */
bool profit_optimizer_should_fertilize(const Agent *self, const FarmState *state,
                                        int planted_index) {
    (void)self;
    const PlantedCrop *planted = &state->planted.data[planted_index];
    const CropDef *crop = config_find_crop(state->config, planted->crop_item_id);
    const FertilizerConfig *fertilizer = &state->config->fertilizer;

    double marginal_profit = economy_fertilizer_expected_marginal_profit(crop, fertilizer, state->config);
    bool low_soil_health = economy_soil_health_factor(state) < SOIL_HEALTH_FERTILIZE_THRESHOLD;
    if (low_soil_health && marginal_profit > SOIL_MAINTENANCE_MARGINAL_PROFIT_FLOOR) {
        return state->money >= fertilizer->cost;
    }
    if (!economy_can_spend_with_reserve(state, fertilizer->cost)) {
        return false;
    }
    return marginal_profit > 0;
}

/* agents/profit_optimizer.py:101-110 */
void profit_optimizer_choose_contracts(const Agent *self, const FarmState *state,
                                        const ResolvedConfig *config, ContractDecisionBuffer *out) {
    (void)self;
    for (size_t i = 0; i < state->active_contracts.count; i++) {
        if (!state->active_contracts.data[i].resolved) {
            return; /* any unresolved active contract -> accept nothing */
        }
    }
    const ContractRecord *best = NULL;
    for (size_t i = 0; i < state->contract_offers.count; i++) {
        const ContractRecord *offer = &state->contract_offers.data[i];
        if (offer->resolved) {
            continue;
        }
        if (contracts_is_offer_profitable(state, config, offer) &&
            contracts_is_offer_feasible(state, config, offer)) {
            if (best == NULL || offer->unit_price > best->unit_price) {
                best = offer;
            }
        }
    }
    if (best != NULL) {
        contract_decision_push(out, best->id);
    }
}

typedef struct {
    double margin;
    const RecipeDef *recipe;
    size_t original_index;
} ScoredRecipe;

/* Stable sort by descending margin (ties keep `recipes`' own order) --
 * decorate-with-index since qsort isn't guaranteed stable. */
static int cmp_scored_recipe(const void *a, const void *b) {
    const ScoredRecipe *sa = a;
    const ScoredRecipe *sb = b;
    if (sa->margin != sb->margin) {
        return sa->margin > sb->margin ? -1 : 1;
    }
    return (sa->original_index > sb->original_index) - (sa->original_index < sb->original_index);
}

/* agents/profit_optimizer.py:112-169 */
void profit_optimizer_choose_processing(const Agent *self, const FarmState *state,
                                         const ResolvedConfig *config,
                                         ProcessingDecisionBuffer *out) {
    (void)self;
    int remaining_capacity =
        (state->has_processing_capacity ? state->processing_capacity : 0) -
        (int)state->processing_jobs.count;
    if (remaining_capacity <= 0) {
        return;
    }
    if (config->recipe_count == 0) {
        return;
    }

    ScoredRecipe *profitable = malloc(config->recipe_count * sizeof(ScoredRecipe));
    size_t profitable_count = 0;
    for (size_t i = 0; i < config->recipe_count; i++) {
        const RecipeDef *recipe = &config->recipes[i];
        double input_price =
            state->has_market_price[recipe->input_item_id] ? state->market_prices[recipe->input_item_id] : 0.0;
        double output_price = state->has_market_price[recipe->output_item_id]
                                   ? state->market_prices[recipe->output_item_id]
                                   : 0.0;
        double input_value = input_price * recipe->input_quantity;
        double output_value = output_price * recipe->output_quantity;
        double margin_per_batch = output_value - input_value - recipe->cost;
        if (margin_per_batch > 0) {
            profitable[profitable_count++] =
                (ScoredRecipe){.margin = margin_per_batch, .recipe = recipe, .original_index = i};
        }
    }
    qsort(profitable, profitable_count, sizeof(ScoredRecipe), cmp_scored_recipe);

    typedef struct {
        ItemId input_item_id;
        int reserved;
    } Reservation;
    Reservation *reserved = malloc((profitable_count > 0 ? profitable_count : 1) * sizeof(Reservation));
    size_t reserved_count = 0;
    double cash_remaining = state->money;

    for (size_t i = 0; i < profitable_count && remaining_capacity > 0; i++) {
        const RecipeDef *recipe = profitable[i].recipe;
        ItemId input_item_id = recipe->input_item_id;

        Reservation *slot = NULL;
        for (size_t r = 0; r < reserved_count; r++) {
            if (reserved[r].input_item_id == input_item_id) {
                slot = &reserved[r];
                break;
            }
        }
        int already_reserved = slot != NULL ? slot->reserved : 0;

        int available =
            inventory_available_quantity(state, input_item_id, recipe->min_quality) - already_reserved;
        int max_by_input = int_floor_div(available, recipe->input_quantity);
        double cost_per_batch = recipe->cost;
        int max_by_cash =
            cost_per_batch > 0 ? (int)(cash_remaining / cost_per_batch) : max_by_input;
        int batches = remaining_capacity;
        if (max_by_input < batches) {
            batches = max_by_input;
        }
        if (max_by_cash < batches) {
            batches = max_by_cash;
        }
        if (batches <= 0) {
            continue;
        }

        processing_decision_push(out, (ProcessingDecision){.recipe_id = recipe->id, .batches = batches});
        remaining_capacity -= batches;
        if (slot == NULL) {
            reserved[reserved_count++] =
                (Reservation){.input_item_id = input_item_id,
                               .reserved = already_reserved + batches * recipe->input_quantity};
        } else {
            slot->reserved = already_reserved + batches * recipe->input_quantity;
        }
        cash_remaining -= batches * cost_per_batch;
    }

    free(reserved);
    free(profitable);
}

/* agents/profit_optimizer.py:171-172 */
void profit_optimizer_choose_sales(const Agent *self, const FarmState *state,
                                    const ResolvedConfig *config, SalesDecisionBuffer *out) {
    (void)self;
    agent_route_sales_by_best_price(state, config, out);
}

/* agents/profit_optimizer.py:174-185 */
bool profit_optimizer_should_use_fertilizer(const Agent *self, const FarmState *state,
                                             ItemId crop_item_id) {
    (void)self;
    const CropDef *crop = config_find_crop(state->config, crop_item_id);
    const FertilizerConfig *fertilizer = &state->config->fertilizer;

    double marginal_profit = economy_fertilizer_expected_marginal_profit(crop, fertilizer, state->config);
    bool low_soil_health = economy_soil_health_factor(state) < SOIL_HEALTH_FERTILIZE_THRESHOLD;
    if (low_soil_health && marginal_profit > SOIL_MAINTENANCE_MARGINAL_PROFIT_FLOOR) {
        return state->money >= crop->seed_cost + fertilizer->cost;
    }
    if (!economy_can_spend_with_reserve(state, crop->seed_cost + fertilizer->cost)) {
        return false;
    }
    return marginal_profit > 0;
}

const Agent AGENT_PROFIT_OPTIMIZER = {
    .name = "profit_optimizer",
    .description = "Maximizes expected profit per slot per day; waters reliably; fertilizes only "
                    "when the math says it pays off.",
    .watering_diligence = 1.0,
    .choose_crop = profit_optimizer_choose_crop,
    .should_buy_upgrade = profit_optimizer_should_buy_upgrade,
    .should_water = agent_base_should_water,
    .should_fertilize = profit_optimizer_should_fertilize,
    .choose_contracts = profit_optimizer_choose_contracts,
    .choose_contract_deliveries = agent_base_choose_contract_deliveries,
    .choose_processing = profit_optimizer_choose_processing,
    .choose_sales = profit_optimizer_choose_sales,
    .should_use_fertilizer = profit_optimizer_should_use_fertilizer,
};
